README.md
## Bio::VertRes::Permissions
Update permissions of files efficiently in parallel.

[![Build Status](https://travis-ci.org/sanger-pathogens/Bio-VertRes-Permissions.svg?branch=master)](https://travis-ci.org/sanger-pathogens/Bio-VertRes-Permissions)   
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-brightgreen.svg)](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/blob/master/GPL-LICENSE)   

## Contents
* [Introduction](#introduction)
* [Installation](#installation)
  * [From source](#from-source)
  * [Running the tests](#running-the-tests)
* [Usage](#usage)
* [License](#license)
* [Feedback/Issues](#feedbackissues)

## Introduction
Take in a list of directories located on multiple disk arrays and update the permissions of the files efficiently in parallel.
The directories are first partitioned into different disk arrays. Multiple arrays are processed in parallel, with multiple threads per array.

## Installation
Details for installing Bio::VertRes::Permissions are provided below. If you encounter an issue when installing Bio::VertRes::Permissions please contact your local system administrator. If you encounter a bug please log it [here](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/issues) or email us at path-help@sanger.ac.uk.

### From source
Clone the repository:   
   
`git clone https://github.com/sanger-pathogens/Bio-VertRes-Permissions.git`   
   
Move into the directory and install all dependencies using [DistZilla](http://dzil.org/):   
  
```
cd Bio-VertRes-Permissions
dzil authordeps --missing | cpanm
dzil listdeps --missing | cpanm
```
  
Run the tests:   
  
`dzil test`   
If the tests pass, install Bio-VertRes-Permissions:   
  
`dzil install`   

### Running the tests
The test can be run with dzil from the top level directory:  
  
`dzil test`  

## Usage
```
Usage: update_pipeline_file_permissions [options]
Changes the permissions of all files in a set of directories

Options: -t STR    type [file_of_directories]
         -i STR    input filename of directories []
         -u STR    username [pathpipe]
         -g STR    unix group [pathogen]
         -o STR    file permissions in octal [0750]
         -l INT    directory level to split on [2]
         -d INT    threads per disk array [1]
         -p INT    num disk arrays to process in parallel [1]
         -v        verbose output to STDOUT
         -h        this help message

Example: Run with defaults
         update_pipeline_file_permissions -i file_of_directories.txt

Example: Update group to team81 in parallel
         bsub.py --threads 16 5 log update_pipeline_file_permissions -i file_of_directories.txt -d 4 -p 4 -g team81
```
## License
Bio::VertRes::Permissions is free software, licensed under [GPLv3](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/blob/master/GPL-LICENSE).

## Feedback/Issues
Please report any issues to the [issues page](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/issues) or email path-help@sanger.ac.uk.


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/tree/HEAD)

**Merged pull requests:**

- Update readme [\#5](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/pull/5) ([ssjunnebo](https://github.com/ssjunnebo))
- Speedup file changes [\#4](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/pull/4) ([andrewjpage](https://github.com/andrewjpage))
- change to dist zilla starter bundle [\#3](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/pull/3) ([nds](https://github.com/nds))
- Allow parallel script [\#2](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/pull/2) ([andrewjpage](https://github.com/andrewjpage))
- Initial script [\#1](https://github.com/sanger-pathogens/Bio-VertRes-Permissions/pull/1) ([andrewjpage](https://github.com/andrewjpage))



lib/Bio/VertRes/Permissions.pm
package Bio::VertRes::Permissions;

# ABSTRACT:  Modify the permissions of all files in an input set of directories

=head1 SYNOPSIS

Modify the permissions of all files in an input set of directories
   use Bio::VertRes::Permissions;
   
   my $obj = Bio::VertRes::Permissions->new(input_directories => \@directories, partition_level => 2, threads => 1, group => 'abc', user => 'efg', octal_permissions => '0700');
   $obj->update_permissions();

=cut

use Moose;
use Bio::VertRes::Permissions::PartitionDirectories;
use Bio::VertRes::Permissions::ModifyPermissions;
use Parallel::ForkManager;
with 'Bio::VertRes::Permissions::LoggerRole';

has 'input_directories'                  => ( is => 'ro', isa => 'ArrayRef', required => 1 );
has 'partition_level'                    => ( is => 'ro', isa => 'Int',      default  => 2 );
has 'threads_per_disk_array'             => ( is => 'ro', isa => 'Int',      default  => 1 );
has 'num_disk_arrays_to_process_at_once' => ( is => 'ro', isa => 'Int',      default  => 1 );
has 'group'                              => ( is => 'ro', isa => 'Str',      required => 0 );
has 'user'                               => ( is => 'ro', isa => 'Str',      required => 0 );
has 'octal_permissions'                  => ( is => 'ro', isa => 'Str',      default  => '0750' );
has '_partitioned_directories'           => ( is => 'ro', isa => 'ArrayRef', lazy => 1, builder => '_build__partitioned_directories' );

sub _build__partitioned_directories {
    my ($self) = @_;
    return Bio::VertRes::Permissions::PartitionDirectories->new( 
	    input_directories => $self->input_directories, 
	    logger            => $self->logger )
      ->partitioned_directories;
}

sub update_permissions {
    my ($self) = @_;

    my $pm = new Parallel::ForkManager( $self->num_disk_arrays_to_process_at_once );
    for my $disk_array_directories ( @{ $self->_partitioned_directories } ) {
        $pm->start and next;

        my $modify_permissions = Bio::VertRes::Permissions::ModifyPermissions->new(
            input_directories => $disk_array_directories,
            user              => $self->user,
            group             => $self->group,
            octal_permissions => $self->octal_permissions,
            threads           => $self->threads_per_disk_array,
            logger            => $self->logger
        );
        $modify_permissions->update_permissions;

        $pm->finish;
    }
    $pm->wait_all_children;
    return $self;

}

no Moose;
__PACKAGE__->meta->make_immutable;

1;


lib/Bio/VertRes/Permissions/Groups.pm
package Bio::VertRes::Permissions::Groups;

# ABSTRACT:  Groups for the current user

=head1 SYNOPSIS

Groups for the current user
   use Bio::VertRes::Permissions::Groups;
   
   my $obj = Bio::VertRes::Permissions::Groups->new();
   $obj->groups();
   $obj->is_member_of_group();

=cut

use Moose;
with 'Bio::VertRes::Permissions::LoggerRole';

has 'groups' => ( is => 'ro', isa => 'ArrayRef', lazy => 1, builder => '_build_groups' );

sub _build_groups {
    my ($self) = @_;
	# Unix::Groups fails to compile on OSX mountain lion
	my $raw_groups_output = `groups`;
	my @groups = split(/[\s\t]/, $raw_groups_output);
	$self->logger->info( "The current users groups are: " . join( ',', @groups ) );

    return \@groups;
}

sub is_member_of_group {
    my ( $self, $input_group ) = @_;
	
	# The input group can be a space separated list of group names, so check all and return when 1 is found.
	my @input_groups = split(/[\s\t]+/,$input_group);
	
	for my $current_group (@input_groups)
	{
        if ( grep { $_ eq $current_group } @{ $self->groups } ) {
	    	$self->logger->info("The current users groups include $input_group");
            return $current_group;
        }
    }
	$self->logger->info("The current users groups do not include $input_group");
    return undef;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;


lib/Bio/VertRes/Permissions/ModifyPermissions.pm
package Bio::VertRes::Permissions::ModifyPermissions;

# ABSTRACT:  Take in a list of directories and update the permissions and owners of all files

=head1 SYNOPSIS

Take in a list of directories and update the permissions and owners of all files
   use Bio::VertRes::Permissions::ModifyPermissions;
   
   my $obj = Bio::VertRes::Permissions::ModifyPermissions->new(input_directories => \@directories, user => 'xyz', group => 'abc', octal_permissions => 0750);
   $obj->update_permissions;


=cut

use Moose;
use Parallel::ForkManager;
use File::Find;
with 'Bio::VertRes::Permissions::LoggerRole';

has 'input_directories' => (
    is       => 'ro',
    isa      => 'ArrayRef',
    required => 1
);
has 'threads'           => ( is => 'ro', isa => 'Int', default  => 1 );
has 'group'             => ( is => 'ro', isa => 'Str', required => 1 );
has 'user'              => ( is => 'ro', isa => 'Str', lazy => 1, builder => '_build_user' );

has '_uid'              => ( is => 'ro', isa => 'Num', lazy => 1, builder => '_build__uid' );
has '_gid'              => ( is => 'ro', isa => 'Num', lazy => 1, builder => '_build__gid' );

has 'octal_permissions' => ( is => 'ro', isa => 'Num', default  => 0750 );

sub BUILD {
    my ($self) = @_;
    $self->logger->info( "Changing file permissions - threads: " . $self->threads );
    $self->logger->info( "Changing file permissions - permissions: " . $self->octal_permissions );
    $self->logger->info( "Changing file permissions - user: " . $self->user );
    $self->logger->info( "Changing file permissions - group: " . $self->group );
}

sub _build__uid
{
	my ($self) = @_;
	my $uid = getpwnam $self->user;
	return $uid;
}

sub _build__gid
{
	my ($self) = @_;
	my $gid = getgrnam $self->group;
	return $gid;
}

sub _build_user
{
	my ($self) = @_;
	return $ENV{LOGNAME} || $ENV{USER} || getpwuid($<);
}

sub _wanted {
    my $self = ${ $_[0] }{self};
    return unless ( defined $File::Find::name );

    my ($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size, $atime,$mtime,$ctime,$blksize,$blocks)= stat($File::Find::name);
    chmod $self->octal_permissions, $File::Find::name;
    if($gid != $self->_gid )
    {
       chown $self->_uid, $self->_gid, $_;
    }
}

sub update_permissions {
    my ($self) = @_;

    my $pm = new Parallel::ForkManager( $self->threads );
    for my $directory ( @{ $self->input_directories } ) {
        $pm->start and next;

        find(
            {
                wanted => sub {
                    _wanted( { self => $self } );
                },
                follow => 1
            },
            $directory
        );
        $pm->finish;
    }
    $pm->wait_all_children;
    return $self;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;


lib/Bio/VertRes/Permissions/PartitionDirectories.pm
package Bio::VertRes::Permissions::PartitionDirectories;

# ABSTRACT:  Split up directories into different disk arrays

=head1 SYNOPSIS

Split up directories into different disk arrays
   use Bio::VertRes::Permissions::PartitionDirectories;
   
   my $obj = Bio::VertRes::Permissions::PartitionDirectories->new(input_directories => \@directories);
   $obj->partitioned_directories();

=cut

use Moose;

has 'input_directories'       => ( is => 'ro', isa => 'ArrayRef', required => 1 );
has 'partition_level'         => ( is => 'ro', isa => 'Int', default  => 2 );
has 'partitioned_directories' => ( is => 'ro', isa => 'ArrayRef', lazy     => 1, builder => '_build_partitioned_directories' );

sub _build_partitioned_directories {
    my ($self) = @_;
	my @partitioned_directories;
	
	my %levels_to_directories;
	for my $directory (@{$self->input_directories})
	{
		my @directory_parts = split('/', $directory);
		if(@directory_parts > $self->partition_level && $directory_parts[($self->partition_level) ] ne '')
		{
			push(@{$levels_to_directories{$directory_parts[($self->partition_level) ]} }, $directory);
		}
		else
		{
			push(@{$levels_to_directories{ungrouped}}, $directory);
		}
	}

	for my $level_name (sort keys %levels_to_directories)
	{
		my @sorted_directories = sort @{$levels_to_directories{$level_name}};
		push(@partitioned_directories, \@sorted_directories);
	}
	
	return \@partitioned_directories;
}


no Moose;
__PACKAGE__->meta->make_immutable;

1;


lib/Bio/VertRes/Permissions/CommandLine/UpdatePipelineFilePermissions.pm
undef $VERSION;

package Bio::VertRes::Permissions::CommandLine::UpdatePipelineFilePermissions;

# ABSTRACT: Update all the permissions of files in a list of directories

=head1 SYNOPSIS

Update all the permissions of files in a list of directories

=cut

use Moose;
use Getopt::Long qw(GetOptionsFromArray);
use Bio::VertRes::Permissions;
use Bio::VertRes::Permissions::ParseFileOfDirectories;
use Bio::VertRes::Permissions::Groups;
with 'Bio::VertRes::Permissions::LoggerRole';

has 'args'                               => ( is => 'ro', isa => 'ArrayRef', required => 1 );
has 'script_name'                        => ( is => 'ro', isa => 'Str',      required => 1 );
has 'help'                               => ( is => 'rw', isa => 'Bool',     default  => 0 );
has 'type'                               => ( is => 'rw', isa => 'Str',      default  => 'file_of_directories' );
has 'type_id'                            => ( is => 'rw', isa => 'Str',      default  => '' );
has 'user'                               => ( is => 'rw', isa => 'Str',      default  => 'pathpipe' );
has 'group'                              => ( is => 'rw', isa => 'Str',      default  => 'pathogen' );
has 'octal_permissions'                  => ( is => 'rw', isa => 'Num',      default  => 0750 );
has 'partition_level'                    => ( is => 'rw', isa => 'Int',      default  => 2 );
has 'threads_per_disk_array'             => ( is => 'rw', isa => 'Int',      default  => 1 );
has 'num_disk_arrays_to_process_at_once' => ( is => 'rw', isa => 'Int',      default  => 1 );
has 'verbose'                            => ( is => 'rw', isa => 'Bool',     default  => 0 );

sub BUILD {
    my ($self) = @_;

    my ( $type, $type_id, $user, $group, $octal_permissions, $partition_level, $threads_per_disk_array, $num_disk_arrays_to_process_at_once,
        $verbose, $help );

    GetOptionsFromArray(
        $self->args,
        't|type=s'                             => \$type,
        'i|type_id=s'                          => \$type_id,
        'u|user=s'                             => \$user,
        'g|group=s'                            => \$group,
        'o|octal_permissions=s'                => \$octal_permissions,
        'l|partition_level=i'                  => \$partition_level,
        'd|threads_per_disk_array=i'             => \$threads_per_disk_array,
        'p|num_disk_arrays_to_process_at_once=i' => \$num_disk_arrays_to_process_at_once,
        'v|verbose'                            => \$verbose,
        'h|help'                               => \$help,
    );

    if ( defined($verbose) ) {
        $self->verbose($verbose);
        $self->logger->level(10000);
    }

    $self->help($help) if ( defined($help) );
    ( !$self->help ) or die $self->usage_text;
	( defined( $type_id) && -e  $type_id ) or die $self->usage_text;

    $self->type($type)                                                             if ( defined($type) );
    $self->type_id($type_id)                                                       if ( defined($type_id) );
    $self->user($user)                                                             if ( defined($user) );
    $self->group($group)                                                           if ( defined($group) );
    $self->octal_permissions(oct($octal_permissions))                                   if ( defined($octal_permissions) );
    $self->partition_level($partition_level)                                       if ( defined($partition_level) );
    $self->threads_per_disk_array($threads_per_disk_array)                         if ( defined($threads_per_disk_array) );
    $self->num_disk_arrays_to_process_at_once($num_disk_arrays_to_process_at_once) if ( defined($num_disk_arrays_to_process_at_once) );
}

sub run {
    my ($self) = @_;

    my $groups_obj = Bio::VertRes::Permissions::Groups->new(logger => $self->logger);
	if(! $groups_obj->is_member_of_group($self->group))
	{
		$self->logger->error("The current user does not have permission to use the input group: ".$self->group);
		die();
	}
    
    my $input_directories =
      Bio::VertRes::Permissions::ParseFileOfDirectories->new( input_file => $self->type_id, logger => $self->logger )->directories();

    my $update_permissions_obj = Bio::VertRes::Permissions->new(
        input_directories                  => $input_directories,
        partition_level                    => $self->partition_level,
        threads_per_disk_array             => $self->threads_per_disk_array,
        num_disk_arrays_to_process_at_once => $self->num_disk_arrays_to_process_at_once,
        group                              => $self->group,
        user                               => $self->user,
        octal_permissions                  => $self->octal_permissions,
        logger                             => $self->logger,
    );
    $update_permissions_obj->update_permissions();
}

sub usage_text {
    my ($self) = @_;

    return <<USAGE;
Usage: update_pipeline_file_permissions [options]
Changes the permissions of all files in a set of directories
  
Options: -t STR    type [file_of_directories]
         -i STR    input filename of directories []
         -u STR    username [pathpipe]
         -g STR    unix group [pathogen]
         -o STR    file permissions in octal [0750]
         -l INT    directory level to split on [2]
         -d INT    threads per disk array [1]
         -p INT    num disk arrays to process in parallel [1]
         -v        verbose output to STDOUT
         -h        this help message
		 
Example: Run with defaults
         update_pipeline_file_permissions -i file_of_directories.txt 

Example: Update group to team81 in parallel
         bsub.py --threads 16 update_pipeline_file_permissions -i file_of_directories.txt -d 4 -p 4 -g team81

USAGE
}

__PACKAGE__->meta->make_immutable;
no Moose;
1;
